File: builder/builder.py
from abc import ABCMeta, abstractmethod
from typing import List


class Builder(metaclass=ABCMeta):
    @abstractmethod
    def make_title(self, title: str):
        pass

    @abstractmethod
    def make_string(self, string: str):
        pass

    @abstractmethod
    def make_items(self, items: List[str]):
        pass

    @abstractmethod
    def close(self):
        pass


class Director(object):
    """Directoryに様々なBuilderを渡して共通処理ができる"""

    def __init__(self, builder: "Builder"):
        self.builder = builder

    def construct(self):
        self.builder.make_title("Greeting!")
        self.builder.make_string("朝から昼にかけて、")
        self.builder.make_items(["おはよう。", "こんにちは。"])
        self.builder.make_string("夜に、")
        self.builder.make_items(["こんばんは。", "おやすみ。", "さようなら。"])
        self.builder.close()


class TextBuilder(Builder):
    def __init__(self):
        self._buffer = ""

    def make_title(self, title: str):
        self._buffer += f"=====\n「{title}」\n\n"

    def make_string(self, string: str):
        self._buffer += f"*{string}\n\n"

    def make_items(self, items: List[str]):
        for item in items:
            self._buffer += f"・{item}\n"
        self._buffer += "\n"

    def close(self):
        self._buffer += "=====\n"

    def get_string(self) -> str:
        return self._buffer

File: builder/builder.go
package builder

import "fmt"

type builder interface {
	makeTitle(string)
	makeString(string)
	makeItems([]string)
	close()
}

type Director struct {
	builder builder
}

func (d *Director) Construct() {
	d.builder.makeTitle("Title")
	d.builder.makeString("String1")
	d.builder.makeItems([]string{"Item1", "Item2"})
	d.builder.makeString("String2")
	d.builder.makeItems([]string{"Item1", "Item2", "Item3"})
	d.builder.close()
}

type TextBuilder struct {
	buffer string
}

func (tb *TextBuilder) makeTitle(title string) {
	tb.buffer += fmt.Sprintf("=====\n「%v」\n\n", title)
}

func (tb *TextBuilder) makeString(str string) {
	tb.buffer += fmt.Sprintf("*%v\n\n", str)
}

func (tb *TextBuilder) makeItems(items []string) {
	for _, item := range items {
		tb.buffer += fmt.Sprintf("・%v\n", item)
	}
}

func (tb *TextBuilder) close() {
	tb.buffer += fmt.Sprintf("=====\n")
}

func (tb *TextBuilder) GetString() string {
	return tb.buffer
}
